frontend/src/app/ml-models/ml-model-form/ml-model-form.component.ts

```diff
@@ -0,0 +1,413 @@
+// Copyright 2023 Google Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Component, OnInit } from '@angular/core';
+import {
+  UntypedFormGroup, UntypedFormBuilder, UntypedFormArray,
+  Validators, ValidatorFn, ValidationErrors, AbstractControl
+} from '@angular/forms';
+import { Router, ActivatedRoute } from '@angular/router';
+import { plainToClass } from 'class-transformer';
+
+import { MlModelsService } from '../shared/ml-models.service';
+import { MlModel, Type, UniqueId, HyperParameter, Feature, Label, Variable, BigQueryDataset, Timespan, Source } from 'app/models/ml-model';
+import { MatOption } from '@angular/material/core';
+
+@Component({
+  selector: 'app-ml-model-form',
+  templateUrl: './ml-model-form.component.html',
+  styleUrls: ['./ml-model-form.component.sass']
+})
+export class MlModelFormComponent implements OnInit {
+
+  mlModelForm: UntypedFormGroup;
+  mlModel: MlModel = new MlModel();
+  state: string = 'loading'; // state has one of values: loading, loaded or error
+  title: string = '';
+  error_message: string = '';
+  uniqueIds: string[];
+  types: string[];
+  variables: Variable[] = [];
+  optionDescriptions: boolean = false;
+
+  constructor(
+    private _fb: UntypedFormBuilder,
+    private mlModelsService: MlModelsService,
+    private router: Router,
+    private route: ActivatedRoute) {
+      this.title = this.router.url.endsWith('new') ? 'New Machine-Learning Model' : 'Edit Machine-Learning Model';
+      this.createForm();
+      this.types = Object.values(Type);
+      this.uniqueIds = Object.values(UniqueId);
+    }
+
+  /**
+   * Create the base form, including general structure, defaults, and validators.
+   */
+  createForm() {
+    this.mlModelForm = this._fb.group({
```

Comment:
Do you expect these regexp patterns to change? if so then maybe a constant can be declared for them.

Reply:
I generally reserve constants for a value that is re-used often (and has the same purpose/meaning) or is unclear. In this case neither really apply. The validation for each of these fields needs to be individually alterable and each pattern is wrapped in a "Validators.pattern" which makes the intended use pretty clear.

Reply:
Understood.